apm-agent-plugins/apm-okhttp-plugin/src/main/java/co/elastic/apm/agent/okhttp/OkHttpClientInstrumentation.java

```diff
@@ -0,0 +1,100 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2019 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.okhttp;
+
+import co.elastic.apm.agent.bci.ElasticApmInstrumentation;
+import co.elastic.apm.agent.http.client.HttpClientHelper;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.TraceContext;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.implementation.bytecode.assign.Assigner;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static co.elastic.apm.agent.http.client.HttpClientHelper.HTTP_CLIENT_SPAN_TYPE_PREFIX;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.returns;
+
+public class OkHttpClientInstrumentation extends ElasticApmInstrumentation {
+
+    private static final String SPAN_TYPE_OK_HTTP_CLIENT = HTTP_CLIENT_SPAN_TYPE_PREFIX + "okhttp";
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    private static void onBeforeExecute( @Advice.FieldValue(value = "originalRequest", typing = Assigner.Typing.DYNAMIC, readOnly = false) @Nullable Object originalRequest,
+                                         @Advice.Local("span") Span span) throws IOException {
+
+        if (tracer == null || tracer.activeSpan() == null) {
+            return;
+        }
+        final AbstractSpan<?> parent = tracer.activeSpan();
+
+        if (originalRequest == null) {
+            return;
+        }
+
+        if (originalRequest instanceof com.squareup.okhttp.Request) {
+            com.squareup.okhttp.Request request = (com.squareup.okhttp.Request) originalRequest;
+            span = HttpClientHelper.startHttpClientSpan(parent, request.method(), request.uri(), request.url().getHost(), SPAN_TYPE_OK_HTTP_CLIENT);
+            originalRequest = ((com.squareup.okhttp.Request) originalRequest).newBuilder().addHeader(TraceContext.TRACE_PARENT_HEADER, span.getTraceContext().getOutgoingTraceParentHeader().toString()).build();
```

Comment:
I imagine this produces garbage and side effects, what about the other solution of adding the header during request build? It will complicate things but will be safer and more performant

Reply:
i tried add header before executing of RequestBuilder.build, but that method called twice, i wrote this in https://github.com/elastic/apm-agent-java/issues/329 .

Reply:
Looks like when you call `org.testcontainers.shaded.okhttp3.Request.Builder#header` method, it invokes the `org.testcontainers.shaded.okhttp3.Headers.Builder#set` method, which seems to replace former value, so shouldn't be a problem if called twice.
I didn't look at version 2, but try to look for the same thing.
Some other options: 
1. look into headers and see if already added
2. try to remove and add
3. flag after first header addition somehow, through request or ThreadLocal for example